proc/threads.c

```diff
@@ -1275,88 +1387,228 @@ int threads_sigpost(process_t *process, thread_t *thread, int sig)
+			_proc_threadSetPriority(lock->owner, curr->priority);
+		}
+
+		hal_spinlockClear(&threads_common.spinlock, &sc);
+
+		if (proc_threadWaitEx(&lock->queue, &lock->spinlock, 0, interruptible, scp) == -EINTR) {
+			/* Recalculate lock owner priority (it might have been inherited from the current thread) */
+			if (lock->owner != NULL) {
+				hal_spinlockSet(&threads_common.spinlock, &sc);
+
+				_proc_threadSetPriority(lock->owner, _proc_threadGetPriority(lock->owner));
+
+				hal_spinlockClear(&threads_common.spinlock, &sc);
+			}
+
 			return -EINTR;
+		}
+	}
+
+	if (lock->queue != NULL) {
+		hal_spinlockSet(&threads_common.spinlock, &sc);
+
+		priority = _proc_lockGetPriority(lock);
+		if (priority < curr->priority) {
+			curr->priority = priority;
+		}
+
+		hal_spinlockClear(&threads_common.spinlock, &sc);
 	}
 
-	lock->v = 0;
+	hal_spinlockSet(&curr->spinlock, &sc);
+
+	LIST_ADD(&curr->locks, lock);
+
+	hal_spinlockClear(&curr->spinlock, &sc);
+	lock->owner = curr;
+
 	return EOK;
 }
 
 
 int proc_lockSet(lock_t *lock)
 {
-	int err;
 	spinlock_ctx_t sc;
+	int err;
 
-	if (!hal_started())
+	if (!hal_started()) {
 		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
+
 	err = _proc_lockSet(lock, 0, &sc);
+
 	hal_spinlockClear(&lock->spinlock, &sc);
+
+	if (err < 0) {
+		threads_put(proc_current());
+	}
+
 	return err;
 }
 
 
 int proc_lockSetInterruptible(lock_t *lock)
 {
-	int err;
 	spinlock_ctx_t sc;
+	int err;
+
+	if (!hal_started()) {
+		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
+
 	err = _proc_lockSet(lock, 1, &sc);
+
 	hal_spinlockClear(&lock->spinlock, &sc);
+
+	if (err < 0) {
+		threads_put(proc_current());
+	}
+
 	return err;
 }
 
 
-int proc_lockTry(lock_t *lock)
+static int _proc_lockTry(lock_t *lock)
 {
+	thread_t *curr = proc_current();
 	spinlock_ctx_t sc;
 
-	int err = EOK;
+	if (lock->owner != NULL) {
+		return -EBUSY;
+	}
+
+	hal_spinlockSet(&curr->spinlock, &sc);
 
-	if (!hal_started())
+	LIST_ADD(&curr->locks, lock);
+	curr->refs++;
+
+	hal_spinlockClear(&curr->spinlock, &sc);
+	lock->owner = curr;
+
+	return EOK;
+}
+
+
+int proc_lockTry(lock_t *lock)
+{
+	spinlock_ctx_t sc;
+	int err;
+
+	if (!hal_started()) {
 		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
-	if (lock->v == 0)
-		err = -EBUSY;
 
-	lock->v = 0;
+	err = _proc_lockTry(lock);
+
 	hal_spinlockClear(&lock->spinlock, &sc);
 
 	return err;
 }
 
 
-int _proc_lockClear(lock_t *lock)
+static int _proc_lockClear(lock_t *lock, thread_t **destroy)
 {
-	lock->v = 1;
-	if (lock->queue == NULL || lock->queue == (void *)-1)
-		return 0;
+	thread_t *owner = lock->owner;
+	spinlock_ctx_t sc;
+	int refs, ret = 0;
 
-	proc_threadWakeup(&lock->queue);
-	return 1;
+	if (owner != NULL) {
+		hal_spinlockSet(&owner->spinlock, &sc);
+
+		LIST_REMOVE(&owner->locks, lock);
+		refs = --owner->refs;
+
+		hal_spinlockClear(&owner->spinlock, &sc);
+
+		lock->owner = NULL;
+		if (lock->queue != NULL) {
```

Comment:
queue can be -1 - if threadWakeup has been performed on an empty queue (feature of sticky condtionals).

Reply:
I think that the lock queue can never be equal -1 but I'm adding the change just in case I've missed something. I've also restored calling threads_common.spinlock proteced version of threadWakeup.